frontend/src/field_with_permission.tsx

```diff
@@ -24,15 +24,46 @@ const useStyles = makeStyles({
   }
 });
 
+/**
+ * Regexp validation of the supported types, email, tel, url, and text, where
+ * empty values are always valid, and any text value is valid.
+ *
+ * @param fieldType
+ * @param value
+ */
+function validateType(fieldType: LimitedInputType, value: string | null): boolean {
+  if (!value) {
+    return true
+  }
+  switch (fieldType) {
+    case 'email':
+      // Validation with RFC5322
+      return !!value.match(/(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])/);
+    case 'url':
+      // validation from urlregex.com
+      return !!value.match(/((([A-Za-z]{3,9}:(?:\/\/)?)(?:[\-;:&=\+\$,\w]+@)?[A-Za-z0-9\.\-]+|(?:www\.|[\-;:&=\+\$,\w]+@)[A-Za-z0-9\.\-]+)((?:\/[\+~%\/\.\w\-_]*)?\??(?:[\-\+=&;%@\.\w_]*)#?(?:[\.\!\/\\\w]*))?)/)
+    case 'tel':
+      // very general match - basically allow anything that starts with
+      // `+<num>(0)` or `0`, followed by any combination of numbers, whitespace
+      // and dashes.
+      return !!value.match(/^(\+?[0-9]{1,3}\(0\)|0)[\s0-9\-]+$/)
+  }
+  return true
```

Comment:
If you support `case 'text'` the type-checker will notify you when you add more types and forget to add them to the switch to make it exhaustive.